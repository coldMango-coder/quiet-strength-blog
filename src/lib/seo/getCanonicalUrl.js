/**
 * Generates canonical URLs for trueallyguide.com pages
 * @param {string} pathname - The pathname from router (e.g., '/blog/post')
 * @returns {string} - The canonical URL
 */
export function getCanonicalUrl(pathname) {
  const baseUrl = 'https://trueallyguide.com';
  
  // Normalize pathname
  let normalizedPath = pathname || '/';
  
  // Remove query parameters and hash
  normalizedPath = normalizedPath.split('?')[0].split('#')[0];
  
  // Remove trailing slash except for root
  if (normalizedPath !== '/' && normalizedPath.endsWith('/')) {
    normalizedPath = normalizedPath.slice(0, -1);
  }
  
  // Convert to lowercase
  normalizedPath = normalizedPath.toLowerCase();
  
  // For root, return exact base URL
  if (normalizedPath === '/') {
    return `${baseUrl}/`;
  }
  
  return `${baseUrl}${normalizedPath}`;
}