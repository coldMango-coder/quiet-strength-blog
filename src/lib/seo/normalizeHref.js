/**
 * Normalizes internal links to ensure consistency with canonical URLs
 * @param {string} href - The href attribute to normalize
 * @returns {string} - The normalized href
 */
export function normalizeHref(href) {
  if (!href || typeof href !== 'string') {
    return href;
  }
  
  // Skip external links, mailto, tel, etc.
  if (href.startsWith('http') && !href.startsWith('http://trueallyguide.com') && !href.startsWith('https://trueallyguide.com') && !href.startsWith('http://www.trueallyguide.com') && !href.startsWith('https://www.trueallyguide.com')) {
    return href;
  }
  
  if (href.startsWith('mailto:') || href.startsWith('tel:') || href.startsWith('#') || href.startsWith('javascript:')) {
    return href;
  }
  
  let normalizedHref = href;
  
  // Convert http to https and remove www for our domain (case insensitive)
  normalizedHref = normalizedHref.replace(/^http:\/\/(www\.)?trueallyguide\.com/i, 'https://trueallyguide.com');
  normalizedHref = normalizedHref.replace(/^https:\/\/www\.trueallyguide\.com/i, 'https://trueallyguide.com');
  
  // Extract path for internal links
  let path = normalizedHref;
  if (normalizedHref.startsWith('https://trueallyguide.com')) {
    path = normalizedHref.replace('https://trueallyguide.com', '') || '/';
  }
  
  // Remove query parameters and hash for normalization (but keep in final URL)
  const [pathPart, queryAndHash] = path.split(/[?#]/);
  let normalizedPath = pathPart;
  
  // Convert to lowercase
  normalizedPath = normalizedPath.toLowerCase();
  
  // Remove trailing slash except for root
  if (normalizedPath !== '/' && normalizedPath.endsWith('/')) {
    normalizedPath = normalizedPath.slice(0, -1);
  }
  
  // Get query and hash parts from the original href
  const queryHashPart = href.includes('?') || href.includes('#') ? 
    href.substring(href.search(/[?#]/)) : '';
  
  // Return full URL if original was a full URL to our domain, otherwise return relative path
  if (href.toLowerCase().startsWith('http://') || href.toLowerCase().startsWith('https://')) {
    return `https://trueallyguide.com${normalizedPath === '/' ? '/' : normalizedPath}${queryHashPart}`;
  }
  
  // For relative paths, just return the normalized path
  return `${normalizedPath === '/' ? '/' : normalizedPath}${queryHashPart}`;
}