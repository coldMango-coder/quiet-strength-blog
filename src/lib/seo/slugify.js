import deburr from 'lodash.deburr';

// Transliterate to ASCII, lowercase, hyphenate, and strip non-url-safe chars
export default function slugify(input) {
  if (!input) return '';
  let s = String(input);
  // Remove diacritics using lodash.deburr
  s = deburr(s);
  // Lowercase
  s = s.toLowerCase();
  // Replace non-word sequences with hyphens
  s = s.replace(/[^a-z0-9]+/g, '-');
  // Trim hyphens
  s = s.replace(/^-+|-+$/g, '');
  return s;
}

