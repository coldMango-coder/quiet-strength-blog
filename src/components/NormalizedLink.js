import React from 'react';
import { Link } from 'react-router-dom';
import { normalizeHref } from '../lib/seo/normalizeHref';

/**
 * A wrapper around React Router's Link component that automatically normalizes hrefs
 * to ensure consistency with canonical URLs
 */
const NormalizedLink = ({ to, href, children, ...props }) => {
  // Support both 'to' (React Router) and 'href' (HTML anchor) props
  const originalHref = to || href;
  
  if (!originalHref) {
    return <a {...props}>{children}</a>;
  }
  
  const normalizedHref = normalizeHref(originalHref);
  
  // Handle null/undefined normalized href
  if (!normalizedHref || typeof normalizedHref !== 'string') {
    return <a {...props}>{children}</a>;
  }
  
  // Check if this is an external link or special link type
  const isExternal = normalizedHref.startsWith('http') && !normalizedHref.startsWith('https://trueallyguide.com');
  const isSpecial = normalizedHref.startsWith('mailto:') || 
                   normalizedHref.startsWith('tel:') || 
                   normalizedHref.startsWith('#') || 
                   normalizedHref.startsWith('javascript:');
  
  // Use regular anchor tag for external links and special links
  if (isExternal || isSpecial) {
    return (
      <a href={normalizedHref} {...props}>
        {children}
      </a>
    );
  }
  
  // Convert full URLs back to relative paths for React Router
  let routerPath = normalizedHref;
  if (normalizedHref.startsWith('https://trueallyguide.com')) {
    routerPath = normalizedHref.replace('https://trueallyguide.com', '') || '/';
  }
  
  // Use React Router Link for internal navigation
  return (
    <Link to={routerPath} {...props}>
      {children}
    </Link>
  );
};

export default NormalizedLink;