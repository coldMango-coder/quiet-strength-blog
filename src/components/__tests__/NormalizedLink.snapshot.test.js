import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';
import NormalizedLink from '../NormalizedLink';
import { getCanonicalUrl } from '../../lib/seo/getCanonicalUrl';

const renderWithRouter = (component) => {
  return render(
    <MemoryRouter>
      {component}
    </MemoryRouter>
  );
};

// Snapshot test for blog post content with internal links
describe('NormalizedLink Internal Links Test', () => {
  test('normalizes internal links to match canonical URLs', () => {
    // Simulate a blog post component with various types of links
    const BlogPostContent = () => (
      <article>
        <h1>Sample Blog Post</h1>
        <p>
          Check out our <NormalizedLink to="/Blog/">blog</NormalizedLink> for more articles.
        </p>
        <p>
          You might also like our article on{' '}
          <NormalizedLink to="/category/Self-Development/">self-development</NormalizedLink>.
        </p>
        <p>
          Read about <NormalizedLink to="https://trueallyguide.com/blog/Introvert-Tips/">introvert tips</NormalizedLink>.
        </p>
        <p>
          Visit our <NormalizedLink to="/">homepage</NormalizedLink> or{' '}
          <NormalizedLink href="mailto:contact@trueallyguide.com">contact us</NormalizedLink>.
        </p>
        <p>
          External reference: <NormalizedLink to="https://example.com">Example</NormalizedLink>
        </p>
        <p>
          Jump to <NormalizedLink to="#section">section</NormalizedLink> below.
        </p>
      </article>
    );

    const { container } = renderWithRouter(<BlogPostContent />);
    
    // Get all internal router links (React Router Links)
    const routerLinks = container.querySelectorAll('a[href]:not([href^="http"]):not([href^="mailto:"]):not([href^="#"])');
    
    // Test that internal links are normalized consistently
    const normalizedLinks = Array.from(routerLinks).map(link => {
      const href = link.getAttribute('href');
      const text = link.textContent;
      const expectedCanonical = getCanonicalUrl(href);
      
      return {
        text,
        actualHref: href,
        expectedCanonical,
        isCorrectlyNormalized: href === expectedCanonical.replace('https://trueallyguide.com', '') || href === expectedCanonical
      };
    });

    // All internal links should be correctly normalized
    const incorrectlyNormalized = normalizedLinks.filter(link => !link.isCorrectlyNormalized);
    
    // Create snapshot of link normalization results
    expect(normalizedLinks).toMatchSnapshot('normalized-internal-links');
    
    // Assert that all internal links are correctly normalized
    expect(incorrectlyNormalized).toHaveLength(0);
    
    // Verify specific expected normalizations
    expect(normalizedLinks).toEqual(expect.arrayContaining([
      expect.objectContaining({ 
        text: 'blog', 
        actualHref: '/blog',
        expectedCanonical: 'https://trueallyguide.com/blog',
        isCorrectlyNormalized: true 
      }),
      expect.objectContaining({ 
        text: 'self-development', 
        actualHref: '/category/self-development',
        expectedCanonical: 'https://trueallyguide.com/category/self-development',
        isCorrectlyNormalized: true 
      }),
      expect.objectContaining({ 
        text: 'introvert tips', 
        actualHref: '/blog/introvert-tips',
        expectedCanonical: 'https://trueallyguide.com/blog/introvert-tips',
        isCorrectlyNormalized: true 
      }),
      expect.objectContaining({ 
        text: 'homepage', 
        actualHref: '/',
        expectedCanonical: 'https://trueallyguide.com/',
        isCorrectlyNormalized: true 
      })
    ]));
    
    // Get external and special links (regular anchor tags)
    const externalLinks = container.querySelectorAll('a[href^="http"], a[href^="mailto:"], a[href^="#"]');
    const externalLinkData = Array.from(externalLinks).map(link => ({
      text: link.textContent,
      href: link.getAttribute('href')
    }));
    
    // Verify external and special links are preserved
    expect(externalLinkData).toEqual(expect.arrayContaining([
      { text: 'contact us', href: 'mailto:contact@trueallyguide.com' },
      { text: 'Example', href: 'https://example.com' },
      { text: 'section', href: '#section' }
    ]));
  });

  test('normalizeHref output matches getCanonicalUrl for internal paths', () => {
    // Test various internal paths
    const testPaths = [
      '/',
      '/blog',
      '/blog/',
      '/Blog/',
      '/category/self-development',
      '/Category/Self-Development/',
      'https://trueallyguide.com/blog/',
      'https://www.trueallyguide.com/Blog/',
      'HTTP://WWW.trueallyguide.com/Category/Test/'
    ];

    testPaths.forEach(path => {
      const { normalizeHref } = require('../../lib/seo/normalizeHref');
      const normalizedHref = normalizeHref(path);
      
      // Extract path part from normalized href
      let pathPart = normalizedHref;
      if (normalizedHref.startsWith('https://trueallyguide.com')) {
        pathPart = normalizedHref.replace('https://trueallyguide.com', '') || '/';
      }
      
      const canonicalUrl = getCanonicalUrl(pathPart);
      
      // The canonical URL should match the normalized href when both are converted to the same format
      const expectedCanonical = canonicalUrl.replace('https://trueallyguide.com', '') || '/';
      
      expect(pathPart).toBe(expectedCanonical);
    });
  });
});