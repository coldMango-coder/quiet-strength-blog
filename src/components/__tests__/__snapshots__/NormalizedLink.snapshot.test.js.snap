// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NormalizedLink Internal Links Test normalizes internal links to match canonical URLs: normalized-internal-links 1`] = `
Array [
  Object {
    "actualHref": "/blog",
    "expectedCanonical": "https://trueallyguide.com/blog",
    "isCorrectlyNormalized": true,
    "text": "blog",
  },
  Object {
    "actualHref": "/category/self-development",
    "expectedCanonical": "https://trueallyguide.com/category/self-development",
    "isCorrectlyNormalized": true,
    "text": "self-development",
  },
  Object {
    "actualHref": "/blog/introvert-tips",
    "expectedCanonical": "https://trueallyguide.com/blog/introvert-tips",
    "isCorrectlyNormalized": true,
    "text": "introvert tips",
  },
  Object {
    "actualHref": "/",
    "expectedCanonical": "https://trueallyguide.com/",
    "isCorrectlyNormalized": true,
    "text": "homepage",
  },
]
`;
